### DISCUSSION TITLE
Why does the NodeJS build take so long?

### DISCUSSION THREAD

**Discussion by nbonbon** [nodejs/help#4591] [Q&A]
Building NodeJS seems to take in between 70-90 minutes.

Comparably things like OpenWRT that is a fully functioning Linux distro take the same amount of time.

Anyone have any input into why NodeJS takes so long to build?

**Comment by F3n67u** [comment]
How do you build node and what is your computer configuration(how many core, memory) is?

There is a lot of code in Node.js including the V8 runtime. I think the mainly cost is from v8.

Build could be speed up by using [Ninja](https://ninja-build.org/). For more information, see [Building Node.js with Ninja](https://github.com/nodejs/node/blob/main/doc/contributing/building-node-with-ninja.md).
Frequent rebuilds could much faster if you config cache, see https://github.com/nodejs/node/blob/main/BUILDING.md#speeding-up-frequent-rebuilds-when-developing.

Related issue: https://github.com/nodejs/help/issues/3015

**Reply by nbonbon** [reply]
Thanks.

Anyone have any input on the potential reduction of time for ninja vs make?

(I did realize that I was forgetting the parallel build option and that did cut it down ... a lot (down to 15') ... but I would still like to do my due diligence and see if I can cut it down more)

**Reply by juanarbol** [reply]
I use ccache, and I had built Node.js literally in 10 seconds.

And yes, V8 is the slower component in terms of compile-time; for that, caching objects had significantly improved the compile-time

**Reply by nbonbon** [reply]
That would only be for incremental builds correct?

**Reply by juanarbol** [reply]
Yes. I think there are some discussions about distributing v8 compiled for cut compile-time, but I'm not sure about it

**Comment by mhdawson** [comment]
ccache is what we use in the Node.js CI to speed things up as well.

**Comment by bnoordhuis** [comment]
> There is a lot of code in Node.js including the V8 runtime. I think the mainly cost is from v8.

That's pretty much the reason. ICU and openssl aren't exactly lightweight either but V8 is the big one. It's very header file and template heavy, meaning the compiler has to parse and analyze 10,000s lines of C++ code again, and again, and again...

At one time V8 supported jumbo/unity builds (#18742) that compiled _much_ faster than regular builds (but needed _lots_ of memory) but maintaining that was a ton of work, so they eventually dropped it.

**Comment by barracuda156** [comment]
> Building NodeJS seems to take in between 70-90 minutes.

I am trying to build NodeJS on a G5, since yesterday :)
90 minutes would have been nice.

**Comment by davidjqq19** [comment]
I've spent 6 hours building it and it's not done yet, in a virtual machine.

**Comment by harshitsaini** [comment]
I generated builds for node using default, ninja and ccache enabled, here is what I found:

command used: `make -j6`
physical cpu cores: 6

- default config: **27 minutes** , incremental builds: **~21 seconds**
- default config + ccache enabled: **52 minutes**, incremental builds:  **~22 seconds**
- ninja: **31 minutes**, incremental builds: **~1 second**
- ninja + ccache enabled: **32 minutes**, incremental builds: **~1 second**

**Comment by umarmughal824** [comment]
I am trying to build it in Ubuntu 23.04 and its still building for over 3 hours.

**Comment by Sur3** [comment]
Yeah somehow every version takes longer and longer to build..

```
     Wed Jan 26 22:05:54 2022 >>> net-libs/nodejs-14.17.6
       merge time: 40 minutes and 25 seconds.

     Sat Jun 11 18:26:32 2022 >>> net-libs/nodejs-16.14.2
       merge time: 1 hour, 10 minutes.

     Mon Jun 13 10:44:19 2022 >>> net-libs/nodejs-16.15.1
       merge time: 1 hour, 49 minutes and 46 seconds.

     Fri Mar 29 05:02:33 2024 >>> net-libs/nodejs-20.11.0
       merge time: 2 hours, 2 minutes and 15 seconds.

     Fri May 31 02:30:24 2024 >>> net-libs/nodejs-20.12.1
       merge time: 2 hours, 48 minutes and 51 seconds.

     Thu Dec 26 18:27:58 2024 >>> net-libs/nodejs-22.4.1-r1
       merge time: 9 hours, 41 minutes and 45 seconds.
```
